File: modules/wem-webapp/src/main/webapp/admin/common/js/content/page/RegionPath.ts
module api.content.page {

    export class RegionPath implements api.Equitable {

        private static DIVIDER = "/";

        private parentComponentPath: ComponentPath;

        private regionName: string;

        private refString: string;

        constructor(parentComponentPath: ComponentPath, regionName: string) {

            this.parentComponentPath = parentComponentPath;
            this.regionName = regionName;
            if (parentComponentPath != null) {
                this.refString = parentComponentPath + "/" + regionName;
            }
            else {
                this.refString = regionName;
            }
        }

        public hasParentComponentPath(): boolean {

            return this.parentComponentPath != null;
        }

        public getParentComponentPath(): ComponentPath {
            return this.parentComponentPath;
        }

        public getRegionName(): string {
            return this.regionName;
        }

        public toString(): string {
            return this.refString;
        }

        equals(o: api.Equitable): boolean {

            if (!(o instanceof RegionPath)) {
                return false;
            }

            var other = <RegionPath>o;

            if (!api.EquitableHelper.stringEquals(this.refString, other.refString)) {
                return false;
            }

            return true;
        }

        public static fromString(str: string): RegionPath {

            var lastDivider = str.lastIndexOf(RegionPath.DIVIDER);
            if (lastDivider == -1) {
                return new RegionPath(null, str);
            }

            var regionNameStart = lastDivider + 1;

            var regionName = str.substring(regionNameStart, str.length);
            var coponentPathAsString = str.substring(0, regionNameStart);
            var parentPath = ComponentPath.fromString(coponentPathAsString);
            return new RegionPath(parentPath, regionName);
        }
    }
}

File: modules/core-api/src/main/java/com/enonic/xp/core/content/page/region/RegionPath.java
package com.enonic.xp.core.content.page.region;

import static com.google.common.base.Preconditions.checkNotNull;
import static org.apache.commons.lang.StringUtils.substringAfterLast;
import static org.apache.commons.lang.StringUtils.substringBeforeLast;

public final class RegionPath
{
    private static final String DIVIDER = "/";

    private final ComponentPath parentComponentPath;

    private final String regionName;

    private final String refString;

    private RegionPath( final ComponentPath parentComponentPath, final String regionName )
    {
        checkNotNull( regionName, "regionName cannot be null" );
        this.parentComponentPath = parentComponentPath;
        this.regionName = regionName;
        if ( parentComponentPath != null )
        {
            this.refString = parentComponentPath + "/" + regionName;
        }
        else
        {
            this.refString = regionName;
        }
    }

    public ComponentPath getParentComponentPath()
    {
        return parentComponentPath;
    }

    public String getRegionName()
    {
        return regionName;
    }

    @Override
    public String toString()
    {
        return refString;
    }

    @Override
    public boolean equals( final Object o )
    {
        if ( this == o )
        {
            return true;
        }
        if ( o == null || !( o instanceof RegionPath ) )
        {
            return false;
        }
        final RegionPath that = (RegionPath) o;
        return refString.equals( that.refString );
    }

    @Override
    public int hashCode()
    {
        return refString.hashCode();
    }

    public static RegionPath from( final String regionPath )
    {
        if ( regionPath.contains( DIVIDER ) )
        {
            final String componentPath = substringBeforeLast( regionPath, DIVIDER );
            final String regionName = substringAfterLast( regionPath, DIVIDER );
            return new RegionPath( ComponentPath.from( componentPath ), regionName );
        }
        else
        {
            return new RegionPath( null, regionPath );
        }
    }

    public static RegionPath from( final ComponentPath parentComponentPath, final String regionName )
    {
        return new RegionPath( parentComponentPath, regionName );
    }
}
